pkg/controller/core/workload_controller.go

```diff
@@ -223,10 +282,11 @@ func (r *WorkloadReconciler) Update(e event.UpdateEvent) bool {
 			log.V(2).Info("ClusterQueue for workload didn't exist; ignored for now")
 		}
 
-	case prevStatus == admitted && status == pending:
+	case prevStatus == admitted && status == cancellingAdmission:
 		if err := r.cache.DeleteWorkload(oldWl); err != nil {
 			log.Error(err, "Failed to delete workload from cache")
 		}
+	case prevStatus == cancellingAdmission && status == pending:
 		// trigger the move of associated inadmissibleWorkloads if required.
 		r.queues.QueueAssociatedInadmissibleWorkloads(ctx, wl)
```

Comment:
Separating these two cases leads to bigger problems #512

Reply:
Any suggestion on how to fix it then?

Reply:
pick one of the transitions to do do both operations. The rest I'm fixing in #512.

Reply:
Interesting, I will need to debug it, but when I naively move the `r.cache.DeleteWorkload` to the other case, then the tests assert that a workload get unadmitted fails. Similarly, if I move requeing to the other case the workload may get readmitted immediately and timeout again (because the WorkloadAdmitted wouldn't be cleaned yet at that point).

Reply:
So the split is what I found working, but I see that there might be atomicity concerns.

Reply:
Ok. I see that I couldn't call `r.cache.DeleteWorkload(oldWl)` in the second case, because at that point the `admission=nil` for the old workload so I don't have reference to the clusterQueue, which is necessary for the `DeleteWorkload` function.

Reply:
I will think how to solve it, but also let me know if you get an idea in the meanwhile, but maybe the `Unknown` solution wasn't that bad after all, let's see.

Reply:
Pushed a proposed fix

Reply:
Done